### ISSUE TITLE
Duplicate entry '1' for key 'PRIMARY' and Kamailio failed to start.

### ISSUE THREAD

**Issue by josemapm** [sipcapture/homer-installer#17]
Hi, I want to try homer with Asterisk.
My scenery:
New Vultr VPS with Debian 8 X64.
I tried with:
bash <( curl -s https://cdn.rawgit.com/sipcapture/homer-installer/master/homer_installer.sh )
and with cd /usr/src -> wget .... -> chmod +x homer_installer.sh -> ./homer_installer.sh
With both the same error.

> Beginning initial data load....
> Starting mysqld
> Starting mysql (via systemctl): mysql.service.
> Creating Databases...
> Creating Tables...
> ERROR 1062 (23000) at line 1: Duplicate entry '1' for key 'PRIMARY'
> Setting root password....
> Patching Homer configuration...
> Restarting apache2 (via systemctl): apache2.service.
> Restarting kamailio (via systemctl): kamailio.serviceJob for kamailio.service failed. See 'systemctl status kamailio.service' and 'journalctl -xn' for details.
>  failed!
> *************************************************************
>       ,;;;;,
>      ;;;;;;;;.     Congratulations! HOMER has been installed!
>    ;;;;;;;;;;;;
>   ;;;;  ;;  ;;;;   <--------------- INVITE ---------------
>   ;;;;  ;;  ;;;;    --------------- 200 OK --------------->
>   ;;;;  ..  ;;;;
>   ;;;;      ;;;;   Your system should be now ready to rock!
>   ;;;;  ;;  ;;;;   Please verify/complete the configuration
>   ,;;;  ;;  ;;;;   files generated by the installer below.
>    ;;;;;;;;;;;;
>     :;;;;;;;;;     THIS SCRIPT IS PROVIDED AS-IS, USE AT
>      ;;;;;;;;      YOUR *OWN* RISK, REVIEW LICENSE & DOCS
> 
> *************************************************************

This is the Kamailio error:
```
● kamailio.service - Kamailio (OpenSER) - the Open Source SIP Server
   Loaded: loaded (/lib/systemd/system/kamailio.service; enabled)
   Active: failed (Result: exit-code) since Sat 2017-03-25 21:05:11 UTC; 12min ago
  Process: 8997 ExecStart=/usr/sbin/kamailio -P /var/run/kamailio/kamailio.pid -f $CFGFILE -m $SHM_MEMORY -M $PKG_MEMORY -u $USER -g $GROUP (code=exited, status=255)
 Main PID: 8740 (code=exited, status=0/SUCCESS)

Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: : <core> [cfg.y:3371]: yyerror_at(): parse error in config file /etc/kamailio/kamailio.cfg, line 108, column 40: Can't set module parameter
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: ERROR: <core> [pvapi.c:828]: pv_parse_spec2(): error searching pvar "hep"
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: ERROR: <core> [pvapi.c:1032]: pv_parse_spec2(): wrong char [0/48] in [$hep(0x00b)] at [5 (5)]
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: : <core> [cfg.y:3368]: yyerror_at(): parse error in config file /etc/kamailio/kamailio.cfg, line 915, column 12-22: Can't get ... $hep(0x00b)
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: ERROR: bad config file (2 errors)
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
Mar 25 21:05:11 homer.oirvelvoz.com kamailio[8997]: loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
Mar 25 21:05:11 homer.oirvelvoz.com systemd[1]: kamailio.service: control process exited, code=exited status=255
Mar 25 21:05:11 homer.oirvelvoz.com systemd[1]: Failed to start Kamailio (OpenSER) - the Open Source SIP Server.
Mar 25 21:05:11 homer.oirvelvoz.com systemd[1]: Unit kamailio.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.
```

After that, I tried with a new server with CentOS 7 and I get the same result.

**Comment by josemapm** [comment]
First I want to apologize because my level of English is not very good, so I hope I can explain correctly.
I've been investigating and found where the error occurs 
`"ERROR 1062 (23000) at line 1: Duplicate entry '1' for key 'PRIMARY'"`

In the install script when the "schema_configuration.sql" is loaded, this sql script insert this 2 rows in the node table
```
INSERT INTO `node` (`id`, `host`, `dbname`, `dbport`, `dbusername`, `dbpassword`, `dbtables`, `name`, `status`) VALUES
(1, '127.0.0.1', 'homer_data', '3306', 'homer_user', 'mysql_password', 'sip_capture', 'homer01', 1),
(21, '10.1.0.7', 'homer_data', '3306', 'homer_user', 'mysql_password', 'sip_capture', 'external', 1);
```
First, I think that is a bad idea to use ID = 21 in the second row, because the ID field is autoincrement!

And then, when the install script tries to execute this line (Line 274 for debian and Line 482 for centos) 
`mysql -u "$DB_USER" -p"$DB_PASS" homer_configuration -e "INSERT INTO node VALUES(1,'mysql','homer_data','3306','"$DB_USER"','"$DB_PASS"','sip_capture','node1', 1);" `

The script shows
`"ERROR 1062 (23000) at line 1: Duplicate entry '1' for key 'PRIMARY'"`

A solution, could be:
Modify the schema_configuration.sql
```
INSERT INTO `node` (`id`, `host`, `dbname`, `dbport`, `dbusername`, `dbpassword`, `dbtables`, `name`, `status`) VALUES
(1, '127.0.0.1', 'homer_data', '3306', 'homer_user', 'mysql_password', 'sip_capture', 'homer01', 1),
(2, '10.1.0.7', 'homer_data', '3306', 'homer_user', 'mysql_password', 'sip_capture', 'external', 1);
```

And modify the install script (lines 274 and 482)
`mysql -u "$DB_USER" -p"$DB_PASS" homer_configuration -e "INSERT INTO node VALUES(3,'mysql','homer_data','3306','"$DB_USER"','"$DB_PASS"','sip_capture','node1', 1);" `

**Comment by zecke** [comment]
> On 25 Mar 2017, at 22:24, josemapm <notifications@github.com> wrote:
> 
> 


> Mar 25 21:05:11 
> homer.oirvelvoz.com
>  kamailio[8997]: : <core> [cfg.y:3371]: yyerror_at(): parse error in config file /etc/kamailio/kamailio.cfg, line 108, column 40: Can't set module parameter
> Mar 25 21:05:11 
> homer.oirvelvoz.com kamailio[8997]: ERROR: <core> [pvapi.c:828]: pv_parse_spec2(): error searching pvar "hep"


Most likely the hep module was not loaded (or isn't availavle)? Can you please verify it?

holger

**Comment by josemapm** [comment]
Thanks for your reply and excuse my ignorance, but how can I check it?

**Comment by 00Asgaroth00** [comment]
I don't see the "hep" module as being listed in the kamailio v4.4 documentation [here](http://kamailio.org/docs/modules/4.4.x/).

Additionally, the sipcapture module in v4.4 does not have a "nonsip_hook" module parameter defined.

I see the same mysql error after using the install script today on CentOS 7 x64, and I also see kamailio failing to start up with the following errors:

```
[root@sipcapture ~]# kamailio -c /etc/kamailio/kamailio.cfg -t
loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
 0(1853) ERROR: <core> [modparam.c:141]: set_mod_param_regex(): parameter <nonsip_hook> of type <2> not found in module <sipcapture>
 0(1853) : <core> [cfg.y:3371]: yyerror_at(): parse error in config file //etc/kamailio/kamailio.cfg, line 108, column 40: Can't set module parameter
 0(1853) ERROR: <core> [pvapi.c:828]: pv_parse_spec2(): error searching pvar "hep"
 0(1853) ERROR: <core> [pvapi.c:1032]: pv_parse_spec2(): wrong char [0/48] in [$hep(0x00b)] at [5 (5)]
 0(1853) : <core> [cfg.y:3368]: yyerror_at(): parse error in config file //etc/kamailio/kamailio.cfg, line 915, column 12-22: Can't get from cache: $hep(0x00b)
ERROR: bad config file (2 errors)
```
nonsip_hook parameter is not defined in the sipcapture module's documentation for v4.4.
the hep pvar must be exported from somewhere but I cannot seem to find where, and the documentation does not reference a hep module anywhere.

**Comment by 00Asgaroth00** [comment]
One option for the auto increment primary key is to pass 0, NULL or the "default" keyword as the value for the auto increment column, that way mysql will automagically just increment the value when inserted, another option is to completely exclude that column in the insert statement.

**Comment by 00Asgaroth00** [comment]
It looks like the installer script copies a v5 kamailio configuration file into a kamailio v4.4 installation. I just made the following change to the installer script to try kamailio v5.0.x:

```
[root@sipcapture ~]# diff -u homer_installer.sh.original homer_installer.sh
--- homer_installer.sh.original 2017-03-28 13:15:56.303234455 +0100
+++ homer_installer.sh  2017-03-28 13:19:45.969662759 +0100
@@ -374,7 +374,8 @@
                        echo "HALT! Something went wrong. Please resolve the errors above and try again."
                        exit 1
                       fi
-               wget http://download.opensuse.org/repositories/home:/kamailio:/v4.4.x-rpms/CentOS_7/home:kamailio:v4.4.x-rpms.repo -O /etc/yum.repos.d/kamailio.repo
+               #wget http://download.opensuse.org/repositories/home:/kamailio:/v4.4.x-rpms/CentOS_7/home:kamailio:v4.4.x-rpms.repo -O /etc/yum.repos.d/kamailio.repo
+               wget http://download.opensuse.org/repositories/home:/kamailio:/v5.0.x-rpms/CentOS_7/home:kamailio:v5.0.x-rpms.repo -O /etc/yum.repos.d/kamailio.repo
           fi
           yum -y update
           yum -y install $COMMON_PKGS libdbi-dbd-mysql perl-DBD-MySQL mysql-community-server kamailio rsyslog kamailio-outbound kamailio-sctp kamailio-tls kamailio-websocket kamailio-jansson kamailio-mysql
```
However, with kamailio v5 installed it also fails to startup:
```
[root@sipcapture ~]# kamailio -V
version: kamailio 5.0.0 (x86_64/linux) 
flags: STATS: Off, USE_TCP, USE_TLS, USE_SCTP, TLS_HOOKS, DISABLE_NAGLE, USE_MCAST, DNS_IP_HACK, SHM_MEM, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, TLSF_MALLOC, DBG_SR_MEMORY, USE_FUTEX, FAST_LOCK-ADAPTIVE_WAIT, USE_DNS_CACHE, USE_DNS_FAILOVER, USE_NAPTR, USE_DST_BLACKLIST, HAVE_RESOLV_RES
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535, DEFAULT PKG_SIZE 8MB
poll method support: poll, epoll_lt, epoll_et, sigio_rt, select.
id: unknown 
compiled on 19:22:29 Mar 16 2017 with gcc 4.8.5
```
with the following errors:
```
Mar 28 13:28:08 sipcapture kamailio: Starting kamailio:
Mar 28 13:28:08 sipcapture kamailio: [  OK  ]
Mar 28 13:28:08 sipcapture systemd: Started SYSV: Kamailio is a fast, reliable and flexible SIP Server..
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: db_mysql [km_my_con.c:124]: db_mysql_new_connection(): driver error: Access denied for user 'homer_user'@'localhost' (using password: YES)
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: <core> [db.c:318]: db_do_init2(): could not add connection to the pool
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: sipcapture [sipcapture.c:1142]: child_init(): unable to connect to database [mysql://homer_user:homer_password@127.0.0.1/homer_data] from capture_mode param.
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: <core> [core/sr_module.c:923]: init_mod_child(): Error while initializing module sipcapture (//usr/lib/x86_64-linux-gnu/kamailio/modules/sipcapture.so)
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: <core> [core/pt.c:330]: fork_process(): init_child failed for process 9, pid 1863, "RTIMER EXEC child=0 timer=ta"
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: rtimer [rtimer_mod.c:169]: child_init(): failed to start timer routine as process
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: <core> [core/sr_module.c:923]: init_mod_child(): Error while initializing module rtimer (//usr/lib/x86_64-linux-gnu/kamailio/modules/rtimer.so)
Mar 28 13:28:08 sipcapture homer[1863]: ERROR: <core> [main.c:1701]: main_loop(): error in init_child
Mar 28 13:28:08 sipcapture homer[1854]: ALERT: <core> [main.c:741]: handle_sigs(): child process 1863 exited normally, status=255
Mar 28 13:28:08 sipcapture kamailio: Stopping kamailio: [FAILED]
```

testing the configuration files yields the following errors, but the test returns with status code 0:
```
[root@sipcapture ~]# kamailio -c /etc/kamailio/kamailio.cfg -t
loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
loading modules under config path: //usr/lib/x86_64-linux-gnu/kamailio/modules/
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x00b
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x00f
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x001
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x003
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x004
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x007
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x008
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x009
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x00a
 0(1909) ERROR: sipcapture [sipcapture.c:2763]: pv_parse_hep_name(): REQUEST, PRE, 0x011
 0(1909) WARNING: <core> [core/socket_info.c:1311]: fix_hostname(): could not rev. resolve 0.0.0.0
 0(1909) WARNING: <core> [core/socket_info.c:1311]: fix_hostname(): could not rev. resolve 0.0.0.0
Listening on 
             udp: 0.0.0.0:9060
             tcp: 0.0.0.0:9061
Aliases: 

config file ok, exiting...
[root@sipcapture ~]# echo $?
0
```

**Comment by 00Asgaroth00** [comment]
It seems the kamailio patching for homer database username/password was not working.

I've updated the homer_installer.sh script for CentOS 7 to:

  [1] Use kamailio v5.0.x opensuse repository
  [2] Use the "default" keyword for the auto-increment primary key (only within the installer script though)
  [3] Properly search/replace the homer db username/password and listen port for kamailio config.
  [4] Properly chown'd the html directories (www-data is used in debian based distro's)

I've attached the patch to this message.


[homer_installer.patch.txt](https://github.com/sipcapture/homer-installer/files/875736/homer_installer.patch.txt)

**Comment by 00Asgaroth00** [comment]
when testing the script for centos 7 and kamailio 4.4, kamailio starts (and fails) with the following error:

```
# kamailio -c /etc/kamailio/kamailio.cfg -t
 0(4857) ERROR: <core> [pvapi.c:828]: pv_parse_spec2(): error searching pvar "hep"
 0(4857) ERROR: <core> [pvapi.c:1032]: pv_parse_spec2(): wrong char [0/48] in [$hep(0x00b)] at [5 (5)]
 0(4857) : <core> [cfg.y:3368]: yyerror_at(): parse error in config file //etc/kamailio/kamailio.cfg, line 916, column 12-22: Can't get from cache: $hep(0x00b)
ERROR: bad config file (1 errors)
```
the hep pvar is not in this version of the sipcapture module.

commenting out the sipcapture:request event route allows kamailio to start up:

for reference, I commented out the following section of kamailio.cfg:
```
event_route[sipcapture:request] {

        xlog("HEP Request!\n");
        xlog("received sipcapture request from $si:$sp\r\n");
        xlog("HEP VERSION $hep(version) request from $si:$sp\r\n");
        xlog("HEP CHUNK Source IP $hep(0x002) request from $si:$sp\r\n");
        #Is it SIP ?
        if($hep(0x00b) == 1){

                #Do parsing internal
                return 1;
        }
        # Save to rtcp data into db partitions
        else if($hep(0x00b) == 5){
                $var(utc) = $utimef(%Y%m%d);
                $var(a) = "rtcp_capture_all_" + $var(utc);
                report_capture("$var(a)", "$hep(0x011)", "$hep(0x00f)");
                drop;
        }
        else
        {
                #If report lets proceed here with payload
                xlog("HEP CHUNK PAYLOAD $hep(0x00f) request from $si:$sp\r\n");
                return 0;
        }
}
```

**Comment by lmangani** [comment]
This is a 5.x config with 4.x and falls into the general issue we're having with the installer... there's a branch for 5.x where you can test and submit any patches.

**Comment by 00Asgaroth00** [comment]
Even installing with the v4 config "homer-config/sipcapture/sipcapture.kamailio" you will notice the sipcapture:request event route at the bottom of the config file which references the $hep pvar which is not available in kamailio v4.4's version of sipcapture.

**Comment by 00Asgaroth00** [comment]
This issue is corrected in pull request #23
